File: PAM_Python/gui.py
import tkinter as tk
from tkinter import filedialog, messagebox
import datetime
from imp_exp import ImpExp
from pam import PAM

class GUI(tk.Tk):
    def __init__(self):
        super().__init__()
        self.title("PAM")
        self.geometry("1020x500")

        self.modImpExp = ImpExp()
        self.modPam = PAM()

        self.init_components()

    def init_components(self):
        # Etiquetas
        tk.Label(self, text="Matriz de disimilitud:").place(x=50, y=50, width=150, height=30)
        tk.Label(self, text="Número de grupos:").place(x=50, y=100, width=150, height=30)
        tk.Label(self, text="Archivo de resultado:").place(x=50, y=150, width=150, height=30)
        tk.Label(self, text="Costo de la solución:").place(x=50, y=200, width=150, height=30)
        tk.Label(self, text="Hora inicial:").place(x=500, y=100, width=100, height=30)
        tk.Label(self, text="Hora final:").place(x=500, y=150, width=100, height=30)

        # Campos de texto
        self.txtFileName = tk.Entry(self)
        self.txtFileName.place(x=200, y=50, width=600, height=30)
        self.txtFileName.config(state='readonly')

        self.txtNClusters = tk.Entry(self)
        self.txtNClusters.place(x=200, y=100, width=100, height=30)

        self.txtFileResult = tk.Entry(self)
        self.txtFileResult.place(x=200, y=150, width=200, height=30)

        self.txtCostoSolucion = tk.Entry(self)
        self.txtCostoSolucion.place(x=200, y=200, width=200, height=30)

        self.txtInitTime = tk.Entry(self)
        self.txtInitTime.place(x=600, y=100, width=200, height=30)

        self.txtFinalTime = tk.Entry(self)
        self.txtFinalTime.place(x=600, y=150, width=200, height=30)

        # Botones
        self.cmdAbrir = tk.Button(self, text="Abrir", command=self.abrir_archivo)
        self.cmdAbrir.place(x=820, y=50, width=100, height=30)

        self.cmdClasificar = tk.Button(self, text="Clasificar", command=self.clasificar)
        self.cmdClasificar.place(x=820, y=100, width=100, height=30)

        self.cmdSalir = tk.Button(self, text="Salir", command=self.destroy)
        self.cmdSalir.place(x=820, y=150, width=100, height=30)

    def abrir_archivo(self):
        file_path = filedialog.askopenfilename()
        if file_path:
            self.txtFileName.config(state='normal')
            self.txtFileName.delete(0, tk.END)
            self.txtFileName.insert(0, file_path)
            self.txtFileName.config(state='readonly')

    def clasificar(self):
        try:
            if not self.txtNClusters.get().strip():
                messagebox.showwarning("Advertencia", "Por favor, ingrese el número de clusters.")
                return

            if not self.txtFileName.get().strip():
                messagebox.showwarning("Advertencia", "Por favor, seleccione un archivo.")
                return

            if not self.txtFileResult.get().strip():
                messagebox.showwarning("Advertencia", "Por favor, ingrese el nombre del archivo de resultado.")
                return

            # Importar la matriz de costos
            self.modImpExp.importMatrixCost(self.txtFileName.get())
            self.modPam.d = self.modImpExp.d
            self.modPam.nObjects = self.modImpExp.nObjects

            # Registrar hora inicial
            current_time = datetime.datetime.now().strftime("%H:%M:%S")
            self.txtInitTime.delete(0, tk.END)
            self.txtInitTime.insert(0, current_time)

            # Ejecutar el algoritmo PAM
            self.modPam.pam(int(self.txtNClusters.get()))
            self.txtCostoSolucion.delete(0, tk.END)
            self.txtCostoSolucion.insert(0, str(self.modPam.dCostoSolucion))

            # Registrar hora final
            current_time = datetime.datetime.now().strftime("%H:%M:%S")
            self.txtFinalTime.delete(0, tk.END)
            self.txtFinalTime.insert(0, current_time)

            # Guardar los resultados en un archivo
            self.send_clusters_to_file(self.txtFileResult.get())

        except ValueError:
            messagebox.showerror("Error", "Error: El número de clusters debe ser un valor entero.")
        except IOError as e:
            messagebox.showerror("Error", f"Error al leer el archivo: {str(e)}")
        except Exception as e:
            messagebox.showerror("Error", f"Error inesperado: {str(e)}")

    def send_clusters_to_file(self, fileName):
        try:
            with open(fileName, 'w') as writer:
                writer.write(f"Costo de la solución encontrada: {self.modPam.dCostoSolucion:.2f}\n")

                for i in range(len(self.modPam.kClusters)):
                    writer.write(f"Cluster no {i + 1}: ")
                    for j in range(len(self.modPam.kClusters[i].items)):
                        obj_name = self.modImpExp.nameObjects[self.modPam.kClusters[i].items[j]]
                        writer.write(f"{obj_name}{('.' if j == len(self.modPam.kClusters[i].items) - 1 else ',')}")
                    writer.write("\n")
                writer.write("_________________________________________________________________")

        except IOError as e:
            messagebox.showerror("Error", f"Error al guardar: {str(e)}")

if __name__ == "__main__":
    app = GUI()
    app.mainloop()

File: PAM_Java/GUI.java
import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.*;
import java.text.SimpleDateFormat;
import java.util.Date;

public class GUI extends JFrame {
    private JTextField txtFileName;
    private JTextField txtNClusters;
    private JTextField txtCostoSolucion;
    private JTextField txtInitTime;
    private JTextField txtFinalTime;
    private JTextField txtFileResult;
    private JButton cmdAbrir;
    private JButton cmdClasificar;
    private JButton cmdSalir;

    private ImpExp modImpExp = new ImpExp();
    private PAM modPam = new PAM();

    public GUI() {
        setTitle("PAM");
        setSize(1020, 500);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLayout(null);

        // Etiquetas
        JLabel lblFileName = new JLabel("Matriz de disimilitud:");
        lblFileName.setBounds(50, 50, 150, 30);
        add(lblFileName);

        JLabel lblNClusters = new JLabel("Número de grupos:");
        lblNClusters.setBounds(50, 100, 150, 30);
        add(lblNClusters);

        JLabel lblFileResult = new JLabel("Archivo de resultado:");
        lblFileResult.setBounds(50, 150, 150, 30);
        add(lblFileResult);

        JLabel lblCostoSolucion = new JLabel("Costo de la solución:");
        lblCostoSolucion.setBounds(50, 200, 150, 30);
        add(lblCostoSolucion);

        JLabel lblInitTime = new JLabel("Hora inicial:");
        lblInitTime.setBounds(500, 100, 100, 30);
        add(lblInitTime);

        JLabel lblFinalTime = new JLabel("Hora final:");
        lblFinalTime.setBounds(500, 150, 100, 30);
        add(lblFinalTime);

        // Campos de texto
        txtFileName = new JTextField();
        txtFileName.setBounds(200, 50, 600, 30);
        txtFileName.setEnabled(false);
        add(txtFileName);

        txtNClusters = new JTextField();
        txtNClusters.setBounds(200, 100, 100, 30);
        add(txtNClusters);

        txtFileResult = new JTextField();
        txtFileResult.setBounds(200, 150, 200, 30);
        add(txtFileResult);

        txtCostoSolucion = new JTextField();
        txtCostoSolucion.setBounds(200, 200, 200, 30);
        add(txtCostoSolucion);

        txtInitTime = new JTextField();
        txtInitTime.setBounds(600, 100, 200, 30);
        add(txtInitTime);

        txtFinalTime = new JTextField();
        txtFinalTime.setBounds(600, 150, 200, 30);
        add(txtFinalTime);

        // Botones
        cmdAbrir = new JButton("Abrir");
        cmdAbrir.setBounds(820, 50, 100, 30);
        add(cmdAbrir);

        cmdClasificar = new JButton("Clasificar");
        cmdClasificar.setBounds(820, 100, 100, 30);
        add(cmdClasificar);

        cmdSalir = new JButton("Salir");
        cmdSalir.setBounds(820, 150, 100, 30);
        add(cmdSalir);

        // Acciones de los botones
        cmdAbrir.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                JFileChooser fileChooser = new JFileChooser();
                if (fileChooser.showOpenDialog(null) == JFileChooser.APPROVE_OPTION) {
                    txtFileName.setText(fileChooser.getSelectedFile().getAbsolutePath());
                }
            }
        });

        cmdClasificar.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                try {
                    if (txtNClusters.getText().trim().isEmpty()) {
                        JOptionPane.showMessageDialog(null, "Por favor, ingrese el número de clusters.");
                        return;
                    }

                    if (txtFileName.getText().trim().isEmpty()) {
                        JOptionPane.showMessageDialog(null, "Por favor, seleccione un archivo.");
                        return;
                    }

                    if (txtFileResult.getText().trim().isEmpty()) {
                        JOptionPane.showMessageDialog(null, "Por favor, ingrese el nombre del archivo de resultado.");
                        return;
                    }

                    // Importar la matriz de costos
                    modImpExp.importMatrixCost(txtFileName.getText());
                    modPam.d = modImpExp.d;
                    modPam.nObjects = modImpExp.nObjects;

                    // Registrar hora inicial
                    txtInitTime.setText(new SimpleDateFormat("HH:mm:ss").format(new Date()));

                    // Ejecutar el algoritmo PAM
                    modPam.pam(Integer.parseInt(txtNClusters.getText()));
                    txtCostoSolucion.setText(String.valueOf(modPam.dCostoSolucion));

                    // Registrar hora final
                    txtFinalTime.setText(new SimpleDateFormat("HH:mm:ss").format(new Date()));

                    // Guardar los resultados en un archivo
                    sendClustersToFile(txtFileResult.getText());

                } catch (NumberFormatException ex) {
                    JOptionPane.showMessageDialog(null, "Error: El número de clusters debe ser un valor entero.");
                } catch (IOException ex) {
                    JOptionPane.showMessageDialog(null, "Error al leer el archivo: " + ex.getMessage());
                } catch (Exception ex) {
                    JOptionPane.showMessageDialog(null, "Error inesperado: " + ex.getMessage());
                }
            }
        });

        cmdSalir.addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent e) {
                System.exit(0);
            }
        });
    }

    private void sendClustersToFile(String fileName) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(fileName))) {
            writer.write("Costo de la solución encontrada: " + String.format("%.2f", modPam.dCostoSolucion));
            writer.newLine();
            
            for (int i = 0; i < modPam.kClusters.length; i++) {
                writer.write("Cluster no " + (i + 1) + ": ");
                for (int j = 0; j < modPam.kClusters[i].items.length; j++) {
                    String objName = modImpExp.nameObjects[modPam.kClusters[i].items[j]];
                    writer.write(objName + (j == modPam.kClusters[i].items.length - 1 ? "." : ","));
                }
                writer.newLine();
            }
            writer.write("_________________________________________________________________");
            
        } catch (IOException ex) {
            JOptionPane.showMessageDialog(null, "Error al guardar: " + ex.getMessage());
        }
    }

    public static void main(String[] args) {
        new GUI().setVisible(true);
    }
}